File: templates/lib/src/http_to_socks_proxy.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io';
import 'dart:typed_data';

import 'log_buffer.dart';

class HttpToSocksProxy {
  HttpToSocksProxy({required this.logBuffer});

  final LogBuffer logBuffer;

  static const _headerDelimiter = <int>[13, 10, 13, 10];
  static const _maxHeaderSize = 64 * 1024;

  ServerSocket? _server;
  final Set<Socket> _clients = <Socket>{};

  Future<void> start({required int listenPort, required int socksPort}) async {
    await stop();

    final server = await ServerSocket.bind(
      InternetAddress.loopbackIPv4,
      listenPort,
      shared: false,
    );
    _server = server;
    logBuffer.add(
      '[mixed-proxy] Listening on 127.0.0.1:$listenPort '
      '-> SOCKS5 127.0.0.1:$socksPort',
    );

    server.listen(
      (client) {
        _clients.add(client);
        unawaited(
          _handleClient(client, socksPort).whenComplete(() {
            _clients.remove(client);
            _closeSocket(client);
          }),
        );
      },
      onError: (Object err) {
        logBuffer.add('[mixed-proxy] Accept error: $err');
      },
      cancelOnError: false,
    );
  }

  Future<void> stop() async {
    final server = _server;
    _server = null;
    if (server != null) {
      await server.close();
    }
    for (final client in _clients.toList()) {
      _closeSocket(client);
    }
    _clients.clear();
  }

  Future<void> _handleClient(Socket client, int socksPort) async {
    final clientReader = _BufferedStreamReader(client);
    _SocksConnection? upstreamHttp;
    Socket? upstreamRaw;
    var isHttpRequest = false;

    try {
      final firstByte = await clientReader.readExact(1);
      final looksLikeHttp = _looksLikeHttpMethodStart(firstByte[0]);
      if (!looksLikeHttp) {
        upstreamRaw = await _connectToSocksServer(socksPort);
        await _tunnel(
          clientStream: clientReader.remainingStream(prefix: firstByte),
          upstreamStream: upstreamRaw,
          clientSocket: client,
          upstreamSocket: upstreamRaw,
        );
        return;
      }

      isHttpRequest = true;
      final remainingHeaders = await clientReader.readUntil(
        _headerDelimiter,
        maxBytes: _maxHeaderSize - 1,
      );
      final requestHeaders = BytesBuilder(copy: false)
        ..add(firstByte)
        ..add(remainingHeaders);
      final parsed = _parseRequest(requestHeaders.takeBytes());

      if (parsed.method == 'CONNECT') {
        final authority = _parseAuthority(parsed.target, defaultPort: 443);
        upstreamHttp = await _connectViaSocks(
          socksPort: socksPort,
          host: authority.host,
          port: authority.port,
        );

        client.add(
          latin1.encode('HTTP/1.1 200 Connection Established\r\n\r\n'),
        );
        await client.flush();
        await _tunnel(
          clientStream: clientReader.remainingStream(),
          upstreamStream: upstreamHttp.reader.remainingStream(),
          clientSocket: client,
          upstreamSocket: upstreamHttp.socket,
        );
      } else {
        final target = _resolveHttpTarget(parsed);
        upstreamHttp = await _connectViaSocks(
          socksPort: socksPort,
          host: target.host,
          port: target.port,
        );

        final rewrittenHeader = _buildForwardRequest(parsed, target);
        await _tunnel(
          clientStream: clientReader.remainingStream(prefix: rewrittenHeader),
          upstreamStream: upstreamHttp.reader.remainingStream(),
          clientSocket: client,
          upstreamSocket: upstreamHttp.socket,
        );
      }
    } catch (err) {
      logBuffer.add('[mixed-proxy] Client error: $err');
      if (isHttpRequest) {
        try {
          client.add(
            utf8.encode(
              'HTTP/1.1 502 Bad Gateway\r\n'
              'Connection: close\r\n'
              'Content-Type: text/plain; charset=utf-8\r\n'
              'Content-Length: 16\r\n\r\n'
              '502 Bad Gateway\n',
            ),
          );
          await client.flush();
        } catch (_) {}
      }
    } finally {
      await clientReader.cancel();
      if (upstreamHttp != null) {
        await upstreamHttp.reader.cancel();
        _closeSocket(upstreamHttp.socket);
      }
      _closeSocket(upstreamRaw);
    }
  }

  Future<void> _tunnel({
    required Stream<Uint8List> clientStream,
    required Stream<Uint8List> upstreamStream,
    required Socket clientSocket,
    required Socket upstreamSocket,
  }) async {
    final upload = _pipe(clientStream, upstreamSocket);
    final download = _pipe(upstreamStream, clientSocket);

    await Future.any([upload, download]);
    _closeSocket(upstreamSocket);
    _closeSocket(clientSocket);
    await Future.wait([upload, download], eagerError: false);
  }

  Future<void> _pipe(Stream<Uint8List> src, Socket dest) async {
    try {
      await dest.addStream(src);
      await dest.flush();
    } catch (_) {}
  }

  bool _looksLikeHttpMethodStart(int value) {
    return (value >= 0x41 && value <= 0x5a) || (value >= 0x61 && value <= 0x7a);
  }

  _ParsedRequest _parseRequest(Uint8List headerBytes) {
    final headerText = latin1.decode(headerBytes, allowInvalid: true);
    final lines = headerText.split('\r\n');
    if (lines.isEmpty || lines.first.trim().isEmpty) {
      throw const FormatException('Invalid proxy request line');
    }

    final parts = lines.first.split(' ');
    if (parts.length < 3) {
      throw const FormatException('Malformed request line');
    }

    final headers = <_HeaderLine>[];
    for (final line in lines.skip(1)) {
      if (line.isEmpty) break;
      final separator = line.indexOf(':');
      if (separator <= 0) continue;
      final name = line.substring(0, separator).trim();
      final value = line.substring(separator + 1).trim();
      headers.add(_HeaderLine(name: name, value: value));
    }

    return _ParsedRequest(
      method: parts[0].toUpperCase(),
      target: parts[1],
      version: parts[2],
      headers: headers,
    );
  }

  _Target _resolveHttpTarget(_ParsedRequest request) {
    final rawTarget = request.target.trim();
    if (rawTarget.startsWith('http://') || rawTarget.startsWith('https://')) {
      final uri = Uri.parse(rawTarget);
      final host = uri.host;
      if (host.isEmpty) {
        throw const FormatException('Missing target host');
      }
      final scheme = uri.scheme.isEmpty ? 'http' : uri.scheme.toLowerCase();
      final port = uri.hasPort ? uri.port : (scheme == 'https' ? 443 : 80);
      final pathAndQuery =
          '${uri.path.isEmpty ? '/' : uri.path}${uri.hasQuery ? '?${uri.query}' : ''}';
      return _Target(
        host: host,
        port: port,
        scheme: scheme,
        pathAndQuery: pathAndQuery,
      );
    }

    final hostHeader = request.headers
        .firstWhere(
          (header) => header.name.toLowerCase() == 'host',
          orElse: () => throw const FormatException('Missing Host header'),
        )
        .value;
    final authority = _parseAuthority(hostHeader, defaultPort: 80);
    final path = rawTarget.startsWith('/') ? rawTarget : '/$rawTarget';
    return _Target(
      host: authority.host,
      port: authority.port,
      scheme: 'http',
      pathAndQuery: path,
    );
  }

  Uint8List _buildForwardRequest(_ParsedRequest request, _Target target) {
    final hostHeader = _formatHostHeader(target);
    final sb = StringBuffer()
      ..write(
        '${request.method} ${target.pathAndQuery} ${request.version}\r\n',
      );

    var hasHost = false;
    for (final header in request.headers) {
      final name = header.name.toLowerCase();
      if (name == 'proxy-connection' ||
          name == 'proxy-authorization' ||
          name == 'connection') {
        continue;
      }
      if (name == 'host') {
        hasHost = true;
        sb.write('Host: $hostHeader\r\n');
        continue;
      }
      sb.write('${header.name}: ${header.value}\r\n');
    }

    if (!hasHost) {
      sb.write('Host: $hostHeader\r\n');
    }
    sb.write('Connection: close\r\n\r\n');
    return Uint8List.fromList(latin1.encode(sb.toString()));
  }

  String _formatHostHeader(_Target target) {
    final defaultPort = target.scheme == 'https' ? 443 : 80;
    final host =
        target.host.contains(':') &&
            !target.host.startsWith('[') &&
            !target.host.endsWith(']')
        ? '[${target.host}]'
        : target.host;
    if (target.port == defaultPort) {
      return host;
    }
    return '$host:${target.port}';
  }

  _Authority _parseAuthority(String authority, {required int defaultPort}) {
    final value = authority.trim();
    if (value.isEmpty) {
      throw const FormatException('Empty authority');
    }

    if (value.startsWith('[')) {
      final end = value.indexOf(']');
      if (end <= 0) {
        throw const FormatException('Invalid IPv6 authority');
      }
      final host = value.substring(1, end);
      if (end + 1 < value.length && value[end + 1] == ':') {
        final portText = value.substring(end + 2);
        final port = int.tryParse(portText);
        if (port == null || port < 1 || port > 65535) {
          throw const FormatException('Invalid authority port');
        }
        return _Authority(host: host, port: port);
      }
      return _Authority(host: host, port: defaultPort);
    }

    final lastColon = value.lastIndexOf(':');
    final hasSingleColon = lastColon > 0 && value.indexOf(':') == lastColon;
    if (hasSingleColon) {
      final host = value.substring(0, lastColon);
      final port = int.tryParse(value.substring(lastColon + 1));
      if (port != null && port >= 1 && port <= 65535) {
        return _Authority(host: host, port: port);
      }
    }

    return _Authority(host: value, port: defaultPort);
  }

  Future<_SocksConnection> _connectViaSocks({
    required int socksPort,
    required String host,
    required int port,
  }) async {
    final socket = await _connectToSocksServer(socksPort);
    final reader = _BufferedStreamReader(socket);

    socket.add(const <int>[0x05, 0x01, 0x00]);
    await socket.flush();
    final greeting = await reader.readExact(2);
    if (greeting.length != 2 || greeting[0] != 0x05 || greeting[1] != 0x00) {
      throw const SocketException('SOCKS5 auth negotiation failed');
    }

    final addressBytes = _buildSocksAddress(host);
    final requestBytes = BytesBuilder(copy: false)
      ..add(const <int>[0x05, 0x01, 0x00])
      ..add(addressBytes)
      ..add(<int>[(port >> 8) & 0xff, port & 0xff]);
    socket.add(requestBytes.takeBytes());
    await socket.flush();

    final head = await reader.readExact(4);
    if (head.length != 4 || head[0] != 0x05 || head[1] != 0x00) {
      throw SocketException(
        'SOCKS5 connect failed (code=${head.length > 1 ? head[1] : -1})',
      );
    }

    final atyp = head[3];
    if (atyp == 0x01) {
      await reader.readExact(6); // IPv4 + port
    } else if (atyp == 0x04) {
      await reader.readExact(18); // IPv6 + port
    } else if (atyp == 0x03) {
      final domainLen = await reader.readExact(1);
      await reader.readExact(domainLen[0] + 2);
    } else {
      throw const SocketException('Unsupported SOCKS5 ATYP');
    }

    return _SocksConnection(socket: socket, reader: reader);
  }

  Future<Socket> _connectToSocksServer(int socksPort) async {
    final socket = await Socket.connect(
      InternetAddress.loopbackIPv4,
      socksPort,
      timeout: const Duration(seconds: 8),
    );
    socket.setOption(SocketOption.tcpNoDelay, true);
    return socket;
  }

  Uint8List _buildSocksAddress(String inputHost) {
    final host = inputHost.trim().replaceAll('[', '').replaceAll(']', '');
    final parsed = InternetAddress.tryParse(host);
    if (parsed != null) {
      if (parsed.type == InternetAddressType.IPv4) {
        return Uint8List.fromList(<int>[0x01, ...parsed.rawAddress]);
      }
      if (parsed.type == InternetAddressType.IPv6) {
        return Uint8List.fromList(<int>[0x04, ...parsed.rawAddress]);
      }
    }

    final domainBytes = utf8.encode(host);
    if (domainBytes.isEmpty || domainBytes.length > 255) {
      throw const FormatException('Invalid target host');
    }
    return Uint8List.fromList(<int>[0x03, domainBytes.length, ...domainBytes]);
  }

  void _closeSocket(Socket? socket) {
    try {
      socket?.destroy();
    } catch (_) {}
  }
}

class _BufferedStreamReader {
  _BufferedStreamReader(Stream<Uint8List> stream)
    : _iterator = StreamIterator<Uint8List>(stream);

  final StreamIterator<Uint8List> _iterator;
  final List<int> _buffer = <int>[];

  Future<Uint8List> readUntil(
    List<int> delimiter, {
    required int maxBytes,
  }) async {
    while (true) {
      final matchIndex = _indexOf(_buffer, delimiter);
      if (matchIndex >= 0) {
        final end = matchIndex + delimiter.length;
        final out = Uint8List.fromList(_buffer.sublist(0, end));
        _buffer.removeRange(0, end);
        return out;
      }
      if (_buffer.length > maxBytes) {
        throw const FormatException('Headers too large');
      }
      if (!await _iterator.moveNext()) {
        throw const SocketException('Unexpected EOF');
      }
      _buffer.addAll(_iterator.current);
    }
  }

  Future<Uint8List> readExact(int length) async {
    while (_buffer.length < length) {
      if (!await _iterator.moveNext()) {
        throw const SocketException('Unexpected EOF');
      }
      _buffer.addAll(_iterator.current);
    }
    final out = Uint8List.fromList(_buffer.sublist(0, length));
    _buffer.removeRange(0, length);
    return out;
  }

  Stream<Uint8List> remainingStream({Uint8List? prefix}) async* {
    if (prefix != null && prefix.isNotEmpty) {
      yield prefix;
    }
    if (_buffer.isNotEmpty) {
      yield Uint8List.fromList(_buffer);
      _buffer.clear();
    }
    while (await _iterator.moveNext()) {
      yield _iterator.current;
    }
  }

  Future<void> cancel() => _iterator.cancel();

  int _indexOf(List<int> data, List<int> needle) {
    if (needle.isEmpty || data.length < needle.length) return -1;
    for (var i = 0; i <= data.length - needle.length; i++) {
      var ok = true;
      for (var j = 0; j < needle.length; j++) {
        if (data[i + j] != needle[j]) {
          ok = false;
          break;
        }
      }
      if (ok) return i;
    }
    return -1;
  }
}

class _HeaderLine {
  _HeaderLine({required this.name, required this.value});

  final String name;
  final String value;
}

class _ParsedRequest {
  _ParsedRequest({
    required this.method,
    required this.target,
    required this.version,
    required this.headers,
  });

  final String method;
  final String target;
  final String version;
  final List<_HeaderLine> headers;
}

class _Target {
  _Target({
    required this.host,
    required this.port,
    required this.scheme,
    required this.pathAndQuery,
  });

  final String host;
  final int port;
  final String scheme;
  final String pathAndQuery;
}

class _Authority {
  _Authority({required this.host, required this.port});

  final String host;
  final int port;
}

class _SocksConnection {
  _SocksConnection({required this.socket, required this.reader});

  final Socket socket;
  final _BufferedStreamReader reader;
}

File: app/android/app/src/main/kotlin/com/pingtunnel/client/app/HttpToSocksProxy.kt
package com.pingtunnel.client.app

import android.util.Log
import java.io.BufferedInputStream
import java.io.ByteArrayOutputStream
import java.io.IOException
import java.io.InputStream
import java.io.OutputStream
import java.net.IDN
import java.net.Inet6Address
import java.net.InetAddress
import java.net.InetSocketAddress
import java.net.ServerSocket
import java.net.Socket
import java.net.SocketException
import java.net.URI
import kotlin.concurrent.thread

class HttpToSocksProxy(
    private val tag: String = "MixedLocalProxy"
) {
    @Volatile
    private var running = false

    @Volatile
    private var serverSocket: ServerSocket? = null

    private val clientSockets = mutableSetOf<Socket>()
    private val clientSocketsLock = Any()

    private data class ParsedRequest(
        val method: String,
        val target: String,
        val version: String,
        val headers: List<Pair<String, String>>
    )

    private data class Target(
        val host: String,
        val port: Int,
        val scheme: String,
        val pathAndQuery: String
    )

    private data class Authority(
        val host: String,
        val port: Int
    )

    fun start(listenPort: Int, socksPort: Int) {
        stop()

        val server = ServerSocket()
        server.reuseAddress = true
        server.bind(InetSocketAddress("127.0.0.1", listenPort))
        serverSocket = server
        running = true

        Log.i(tag, "Listening on 127.0.0.1:$listenPort and forwarding via SOCKS5 127.0.0.1:$socksPort")

        thread(start = true, name = "http-socks-accept") {
            acceptLoop(server, socksPort)
        }
    }

    fun stop() {
        running = false

        try {
            serverSocket?.close()
        } catch (_: Exception) {
        }
        serverSocket = null

        synchronized(clientSocketsLock) {
            clientSockets.toList().forEach { closeQuietly(it) }
            clientSockets.clear()
        }
    }

    private fun acceptLoop(server: ServerSocket, socksPort: Int) {
        while (running) {
            val client = try {
                server.accept()
            } catch (e: SocketException) {
                if (running) {
                    Log.w(tag, "Accept failed: ${e.message}")
                }
                break
            } catch (e: IOException) {
                if (running) {
                    Log.e(tag, "Accept IO error", e)
                }
                break
            }

            addClientSocket(client)
            thread(start = true, name = "http-socks-client") {
                handleClient(client, socksPort)
            }
        }
    }

    private fun handleClient(client: Socket, socksPort: Int) {
        client.tcpNoDelay = true
        var upstream: Socket? = null
        var isHttpRequest = false

        try {
            val clientInput = BufferedInputStream(client.getInputStream())
            val clientOutput = client.getOutputStream()
            val firstByte = clientInput.read()
            if (firstByte == -1) {
                return
            }

            if (!isLikelyHttpMethodStart(firstByte)) {
                upstream = connectToSocksServer(socksPort)
                val upstreamOutput = upstream.getOutputStream()
                upstreamOutput.write(firstByte)
                upstreamOutput.flush()
                tunnel(clientInput, clientOutput, upstream)
                return
            }

            isHttpRequest = true
            val headerBytes = readRequestHeaders(clientInput, firstByte)
            val request = parseRequest(headerBytes)

            if (request.method == "CONNECT") {
                val authority = parseAuthority(request.target, defaultPort = 443)
                upstream = connectViaSocks(
                    socksPort = socksPort,
                    targetHost = authority.host,
                    targetPort = authority.port
                )
                clientOutput.write("HTTP/1.1 200 Connection Established\r\n\r\n".toByteArray(Charsets.ISO_8859_1))
                clientOutput.flush()
                tunnel(clientInput, clientOutput, upstream)
            } else {
                val target = resolveHttpTarget(request)
                upstream = connectViaSocks(
                    socksPort = socksPort,
                    targetHost = target.host,
                    targetPort = target.port
                )

                val rewritten = buildForwardRequest(request, target)
                val upstreamOutput = upstream.getOutputStream()
                upstreamOutput.write(rewritten)
                upstreamOutput.flush()

                tunnel(clientInput, clientOutput, upstream)
            }
        } catch (e: Exception) {
            Log.w(tag, "Client handling failed: ${e.message}")
            if (isHttpRequest) {
                try {
                    writeHttpError(client.getOutputStream(), 502, "Bad Gateway")
                } catch (_: Exception) {
                }
            }
        } finally {
            closeQuietly(upstream)
            removeClientSocket(client)
            closeQuietly(client)
        }
    }

    private fun tunnel(
        clientInput: InputStream,
        clientOutput: OutputStream,
        upstream: Socket
    ) {
        val upstreamInput = upstream.getInputStream()
        val upstreamOutput = upstream.getOutputStream()

        val uploadThread = thread(start = true, name = "http-socks-upload") {
            try {
                copyStream(clientInput, upstreamOutput)
            } catch (_: Exception) {
            } finally {
                try {
                    upstream.shutdownOutput()
                } catch (_: Exception) {
                }
            }
        }

        try {
            copyStream(upstreamInput, clientOutput)
        } finally {
            try {
                clientOutput.flush()
            } catch (_: Exception) {
            }
            try {
                uploadThread.join(300)
            } catch (_: Exception) {
            }
        }
    }

    private fun readRequestHeaders(input: BufferedInputStream, firstByte: Int): ByteArray {
        val out = ByteArrayOutputStream()
        out.write(firstByte)
        var state = if (firstByte == '\r'.code) 1 else 0

        while (out.size() < 64 * 1024) {
            val value = input.read()
            if (value == -1) {
                throw IOException("Client closed before sending request headers")
            }

            out.write(value)
            state = when (state) {
                0 -> if (value == '\r'.code) 1 else 0
                1 -> if (value == '\n'.code) 2 else 0
                2 -> if (value == '\r'.code) 3 else 0
                3 -> if (value == '\n'.code) 4 else 0
                else -> 0
            }

            if (state == 4) {
                return out.toByteArray()
            }
        }

        throw IOException("Request headers too large")
    }

    private fun parseRequest(headers: ByteArray): ParsedRequest {
        val text = headers.toString(Charsets.ISO_8859_1)
        val lines = text.split("\r\n")
        val requestLine = lines.firstOrNull()?.trim().orEmpty()
        if (requestLine.isEmpty()) {
            throw IOException("Empty request line")
        }

        val parts = requestLine.split(" ")
        if (parts.size < 3) {
            throw IOException("Invalid request line: $requestLine")
        }

        val parsedHeaders = mutableListOf<Pair<String, String>>()
        for (line in lines.drop(1)) {
            if (line.isEmpty()) {
                break
            }
            val separator = line.indexOf(':')
            if (separator <= 0) {
                continue
            }
            val name = line.substring(0, separator).trim()
            val value = line.substring(separator + 1).trim()
            parsedHeaders.add(name to value)
        }

        return ParsedRequest(
            method = parts[0].uppercase(),
            target = parts[1],
            version = parts[2],
            headers = parsedHeaders
        )
    }

    private fun resolveHttpTarget(request: ParsedRequest): Target {
        val rawTarget = request.target.trim()
        if (rawTarget.startsWith("http://", ignoreCase = true) ||
            rawTarget.startsWith("https://", ignoreCase = true)
        ) {
            val uri = URI(rawTarget)
            val scheme = (uri.scheme ?: "http").lowercase()
            val host = uri.host ?: throw IOException("Missing target host")
            val port = if (uri.port != -1) uri.port else if (scheme == "https") 443 else 80
            return Target(
                host = host,
                port = port,
                scheme = scheme,
                pathAndQuery = buildPathAndQuery(uri.rawPath, uri.rawQuery)
            )
        }

        val hostHeader = request.headers
            .firstOrNull { it.first.equals("Host", ignoreCase = true) }
            ?.second
            ?: throw IOException("Missing Host header")

        val authority = parseAuthority(hostHeader, defaultPort = 80)
        val path = if (rawTarget.startsWith("/")) rawTarget else "/$rawTarget"
        return Target(
            host = authority.host,
            port = authority.port,
            scheme = "http",
            pathAndQuery = path
        )
    }

    private fun buildForwardRequest(request: ParsedRequest, target: Target): ByteArray {
        val sb = StringBuilder()
        sb.append(request.method)
            .append(' ')
            .append(target.pathAndQuery)
            .append(' ')
            .append(request.version)
            .append("\r\n")

        var hasHost = false
        request.headers.forEach { (name, value) ->
            val lowerName = name.lowercase()
            if (lowerName == "proxy-connection" || lowerName == "proxy-authorization") {
                return@forEach
            }
            if (lowerName == "connection") {
                return@forEach
            }
            if (lowerName == "host") {
                hasHost = true
                sb.append("Host: ")
                    .append(formatHostHeader(target.host, target.port, target.scheme))
                    .append("\r\n")
                return@forEach
            }
            sb.append(name).append(": ").append(value).append("\r\n")
        }

        if (!hasHost) {
            sb.append("Host: ")
                .append(formatHostHeader(target.host, target.port, target.scheme))
                .append("\r\n")
        }

        sb.append("Connection: close\r\n")
        sb.append("\r\n")
        return sb.toString().toByteArray(Charsets.ISO_8859_1)
    }

    private fun parseAuthority(authority: String, defaultPort: Int): Authority {
        val value = authority.trim()
        if (value.isEmpty()) {
            throw IOException("Empty authority")
        }

        if (value.startsWith("[")) {
            val end = value.indexOf(']')
            if (end <= 0) {
                throw IOException("Invalid IPv6 authority: $authority")
            }
            val host = value.substring(1, end)
            val hasPort = end + 1 < value.length && value[end + 1] == ':'
            val port = if (hasPort) {
                value.substring(end + 2).toIntOrNull() ?: throw IOException("Invalid port: $authority")
            } else {
                defaultPort
            }
            return Authority(host, port)
        }

        val lastColon = value.lastIndexOf(':')
        val singleColon = value.indexOf(':') == lastColon
        if (lastColon > 0 && singleColon) {
            val host = value.substring(0, lastColon)
            val port = value.substring(lastColon + 1).toIntOrNull()
                ?: throw IOException("Invalid port: $authority")
            return Authority(host, port)
        }

        return Authority(value, defaultPort)
    }

    private fun formatHostHeader(host: String, port: Int, scheme: String): String {
        val defaultPort = if (scheme == "https") 443 else 80
        val printableHost = if (host.contains(':') && !host.startsWith("[") && !host.endsWith("]")) {
            "[$host]"
        } else {
            host
        }
        return if (port == defaultPort) {
            printableHost
        } else {
            "$printableHost:$port"
        }
    }

    private fun buildPathAndQuery(path: String?, query: String?): String {
        val rawPath = if (path.isNullOrEmpty()) "/" else path
        return if (query.isNullOrEmpty()) rawPath else "$rawPath?$query"
    }

    private fun connectViaSocks(socksPort: Int, targetHost: String, targetPort: Int): Socket {
        val socket = connectToSocksServer(socksPort)

        val input = socket.getInputStream()
        val output = socket.getOutputStream()

        output.write(byteArrayOf(0x05, 0x01, 0x00))
        output.flush()

        val greeting = readExact(input, 2)
        if (greeting[0].toInt() != 0x05 || greeting[1].toInt() != 0x00) {
            throw IOException("SOCKS5 auth negotiation failed")
        }

        val request = ByteArrayOutputStream()
        request.write(byteArrayOf(0x05, 0x01, 0x00))
        request.write(buildSocksAddress(targetHost))
        request.write(byteArrayOf(((targetPort shr 8) and 0xff).toByte(), (targetPort and 0xff).toByte()))

        output.write(request.toByteArray())
        output.flush()

        val responseHead = readExact(input, 4)
        if (responseHead[0].toInt() != 0x05 || responseHead[1].toInt() != 0x00) {
            throw IOException("SOCKS5 connect failed (code=${responseHead[1].toInt() and 0xff})")
        }

        when (responseHead[3].toInt() and 0xff) {
            0x01 -> readExact(input, 4 + 2) // IPv4 + port
            0x04 -> readExact(input, 16 + 2) // IPv6 + port
            0x03 -> {
                val domainLength = readExact(input, 1)[0].toInt() and 0xff
                readExact(input, domainLength + 2)
            }
            else -> throw IOException("Unsupported SOCKS5 address type in response")
        }

        return socket
    }

    private fun connectToSocksServer(socksPort: Int): Socket {
        val socket = Socket()
        socket.tcpNoDelay = true
        socket.connect(InetSocketAddress("127.0.0.1", socksPort), 8000)
        return socket
    }

    private fun isLikelyHttpMethodStart(value: Int): Boolean {
        return (value in 'A'.code..'Z'.code) || (value in 'a'.code..'z'.code)
    }

    private fun buildSocksAddress(hostInput: String): ByteArray {
        val host = hostInput.trim().removePrefix("[").removeSuffix("]")

        val ipv4Parts = host.split('.')
        if (ipv4Parts.size == 4 && ipv4Parts.all { it.toIntOrNull() in 0..255 }) {
            val out = ByteArrayOutputStream()
            out.write(0x01)
            ipv4Parts.forEach { out.write(it.toInt()) }
            return out.toByteArray()
        }

        if (host.contains(':')) {
            val addr = InetAddress.getByName(host)
            if (addr is Inet6Address) {
                val out = ByteArrayOutputStream()
                out.write(0x04)
                out.write(addr.address)
                return out.toByteArray()
            }
        }

        val asciiHost = IDN.toASCII(host)
        val encoded = asciiHost.toByteArray(Charsets.UTF_8)
        if (encoded.isEmpty() || encoded.size > 255) {
            throw IOException("Invalid target host")
        }

        val out = ByteArrayOutputStream()
        out.write(0x03)
        out.write(encoded.size)
        out.write(encoded)
        return out.toByteArray()
    }

    private fun copyStream(input: InputStream, output: OutputStream) {
        val buffer = ByteArray(16 * 1024)
        while (true) {
            val read = input.read(buffer)
            if (read <= 0) {
                break
            }
            output.write(buffer, 0, read)
            output.flush()
        }
    }

    private fun readExact(input: InputStream, length: Int): ByteArray {
        val out = ByteArray(length)
        var offset = 0
        while (offset < length) {
            val read = input.read(out, offset, length - offset)
            if (read == -1) {
                throw IOException("Unexpected EOF")
            }
            offset += read
        }
        return out
    }

    private fun writeHttpError(output: OutputStream, code: Int, message: String) {
        val body = "$code $message\n".toByteArray(Charsets.UTF_8)
        val header = buildString {
            append("HTTP/1.1 ").append(code).append(' ').append(message).append("\r\n")
            append("Connection: close\r\n")
            append("Content-Type: text/plain; charset=utf-8\r\n")
            append("Content-Length: ").append(body.size).append("\r\n")
            append("\r\n")
        }.toByteArray(Charsets.ISO_8859_1)
        output.write(header)
        output.write(body)
        output.flush()
    }

    private fun addClientSocket(socket: Socket) {
        synchronized(clientSocketsLock) {
            clientSockets.add(socket)
        }
    }

    private fun removeClientSocket(socket: Socket) {
        synchronized(clientSocketsLock) {
            clientSockets.remove(socket)
        }
    }

    private fun closeQuietly(socket: Socket?) {
        try {
            socket?.close()
        } catch (_: Exception) {
        }
    }
}
